File: nostr-knowledge/examples/basics/generate-keys/generate_keys.go
/*
NOSTR Key Generation — Go

Generates a secp256k1 keypair for use with the NOSTR protocol.
The private key is used to sign events; the public key is your identity.

Dependencies:
	go get github.com/nbd-wtf/go-nostr

Run:
	go run generate_keys.go

References:
	- NIP-01: https://github.com/nostr-protocol/nips/blob/master/01.md
	- NIP-19: https://github.com/nostr-protocol/nips/blob/master/19.md
*/

package main

import (
	"fmt"
	"log"

	"github.com/nbd-wtf/go-nostr"
	"github.com/nbd-wtf/go-nostr/nip19"
)

func main() {
	// Step 1: Generate a random secret key.
	// go-nostr returns the key as a hex string.
	secretKey := nostr.GeneratePrivateKey()

	// Step 2: Derive the public key from the secret key.
	// This performs elliptic curve multiplication on secp256k1.
	// The result is the x-only public key (32 bytes, hex-encoded).
	publicKey, err := nostr.GetPublicKey(secretKey)
	if err != nil {
		log.Fatalf("Failed to derive public key: %v", err)
	}

	// Step 3: Encode to bech32 format (NIP-19).
	// npub = bech32-encoded public key (starts with "npub1")
	// nsec = bech32-encoded secret key (starts with "nsec1")
	npub, err := nip19.EncodePublicKey(publicKey)
	if err != nil {
		log.Fatalf("Failed to encode npub: %v", err)
	}

	nsec, err := nip19.EncodePrivateKey(secretKey)
	if err != nil {
		log.Fatalf("Failed to encode nsec: %v", err)
	}

	// Step 4: Display the results.
	fmt.Println("=== NOSTR Key Generation ===")
	fmt.Println()
	fmt.Println("Secret key (hex):", secretKey)
	fmt.Println("Public key (hex):", publicKey)
	fmt.Println()
	fmt.Println("npub (bech32):", npub)
	fmt.Println("nsec (bech32):", nsec)
	fmt.Println()
	fmt.Println("WARNING: Never share your secret key (nsec) with anyone.")
	fmt.Println("Your public key (npub) is your identity — share it freely.")
}

File: nostr-knowledge/examples/basics/generate-keys/generate_keys.ts
/**
 * NOSTR Key Generation — TypeScript
 *
 * Generates a secp256k1 keypair for use with the NOSTR protocol.
 * The private key is used to sign events; the public key is your identity.
 *
 * Dependencies:
 *   npm install nostr-tools
 *
 * Run:
 *   npx ts-node generate_keys.ts
 *
 * References:
 *   - NIP-01: https://github.com/nostr-protocol/nips/blob/master/01.md
 *   - NIP-19: https://github.com/nostr-protocol/nips/blob/master/19.md
 */

import { generateSecretKey, getPublicKey } from "nostr-tools/pure";
import { npubEncode, nsecEncode } from "nostr-tools/nip19";

// Step 1: Generate a random 32-byte secret key.
// This uses crypto.getRandomValues under the hood.
const secretKey: Uint8Array = generateSecretKey();

// Step 2: Derive the public key from the secret key.
// This performs elliptic curve multiplication on secp256k1.
// The result is the x-coordinate of the public key point (32 bytes, hex-encoded).
const publicKey: string = getPublicKey(secretKey);

// Step 3: Encode to bech32 format (NIP-19).
// npub = bech32-encoded public key (human-readable, starts with "npub1")
// nsec = bech32-encoded secret key (human-readable, starts with "nsec1")
const npub: string = npubEncode(publicKey);
const nsec: string = nsecEncode(secretKey);

// Step 4: Display the results.
console.log("=== NOSTR Key Generation ===\n");
console.log("Secret key (hex):", Buffer.from(secretKey).toString("hex"));
console.log("Public key (hex):", publicKey);
console.log("");
console.log("npub (bech32):", npub);
console.log("nsec (bech32):", nsec);
console.log("");
console.log("WARNING: Never share your secret key (nsec) with anyone.");
console.log("Your public key (npub) is your identity — share it freely.");
